woocommerce/google-listings-and-ads
# Upgrade to use Node.js 20 and bump npm dependencies

### Changes proposed in this Pull Request:

Closes #2002

This PR:
- Bump npm dependencies and upgrade to use Node.js v20.
- Fix the build config of Fast Refresh mode.
- Address Sass deprecation warnings.
   - Ref: https://sass-lang.com/documentation/breaking-changes/mixed-decls/
- Turn off a few eslint rules:
   - `react/react-in-jsx-scope` rule
   - `jsdoc/check-line-alignment` rule
   - `import/no-named-as-default` rule
- Resolve eslint errors: (💡 It would be easier to review the changes by individual commits)
- Automatically fixed errors by running `npm run lint:js -- --fix`: [c41ef64](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/c41ef64f1b2a3e3cef675729185d8797fded520b)
- `@wordpress/i18n-text-domain` rule: [cdd9aab](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/cdd9aabfd374206cea3b525875f4733ccf95c8b7)
- `jsdoc/no-undefined-types` rule: [67ef0e2](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/67ef0e2225dd452423708c6c62a23946d8bb0280)
- `import/no-unresolved` rule: [c4db2a3](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/c4db2a3f4dd55075bc06d70e0b98beb5eec2a1b7)
- `no-redeclare` rule: [6997b6f](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/6997b6fa2730e0be5c507e2f556710c4d3c8c066)
- `jsdoc/require-yields-check` rule: [47d1f9f](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/47d1f9f44fe466171456a131d0ec74290d1c7d6f)
- `@typescript-eslint/no-use-before-define` rule: [9886344](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/9886344a7fea97645b420481271d0e5a8952a128)
- `jsdoc/tag-lines` rule: [fccc0cd](https://github.com/woocommerce/google-listings-and-ads/pull/2517/commits/fccc0cde1a9e21f1ddd9de9c5805895b7b080eae)

❗ Since the jest testing requires a lot of adjustments, it will be handled by #2544. Two of the eslint rule errors are highly related to jest and are therefore handled by #2544 as well.

### Detailed test instructions:

1. Set Node.js to v20
2. Run `npm install` to see if it can finish without errors
3. Check if the following scripts work well
   - `npm run build`
   - `npm run dev`
   - `npm run start`
   - `npm run start:hot`
   - `npm run lint:css`
   - `npm run lint:js`
      - There are two pre-existing warnings that still can not be fixed after this upgrade
      - Errors from the `testing-library/no-unnecessary-act` and `testing-library/no-unnecessary-act` rules are handled separately in #2544
4. Run `node ./node_modules/.bin/bundlewatch` to check if the check works
5. Do some smoke testing

### Changelog entry

> Dev - Upgrade to use Node.js 20 and bump npm dependencies.
